allauth/socialaccount/providers/apple/views.py

```diff
@@ -0,0 +1,180 @@
+        apple_public_key = self._get_apple_public_key(kid=kid)
+
+        public_key = jwt.algorithms.RSAAlgorithm.from_jwk(json.dumps(apple_public_key))
+        return public_key
+
+    def get_client_id(self, provider):
+        app = SocialApp.objects.get(provider=provider.id)
+        return [aud.strip() for aud in app.client_id.split(",")]
+
+    def get_verified_identity_data(self, id_token):
+        provider = self.get_provider()
+        allowed_auds = self.get_client_id(provider)
+
+        try:
+            public_key = self.get_public_key(id_token)
+            identity_data = jwt.decode(
+                id_token,
+                public_key,
+                algorithms=["RS256"],
+                verify=True,
+                audience=allowed_auds,
+                issuer="https://appleid.apple.com",
+            )
+            return identity_data
+
+        except jwt.PyJWTError as e:
+            raise OAuth2Error("Invalid id_token") from e
+
+    def parse_token(self, data):
+        token = SocialToken(
+            token=data["access_token"],
+        )
+        token.token_secret=data.get("refresh_token", "")
+
+        expires_in = data.get(self.expires_in_key)
+        if expires_in:
+            token.expires_at = (
+                timezone.now() + timedelta(seconds=int(expires_in))
+            )
+
+        # `user_data` is a big flat dictionary with the parsed JWT claims
+        # access_tokens, and user info from the apple post.
+        identity_data = self.get_verified_identity_data(data["id_token"])
+        token.user_data = {**data, **identity_data}
+
+        return token
+
+    def complete_login(self, request, app, token, **kwargs):
+        extra_data = token.user_data
+        login = self.get_provider().sociallogin_from_response(
+            request=request, response=extra_data
+        )
+        login.state["id_token"] = token.user_data
+
+        # We can safely remove the apple login session now
+        # Note: The cookie will remain, but it's set to delete on browser close
+        try:
+            request.apple_login_session.delete()
+        except AttributeError:
+            pass
+
+        return login
+
+    def get_user_scope_data(self, request):
+        user_scope_data = request.apple_login_session.get("user", "")
+        try:
+            return json.loads(user_scope_data)
+        except json.JSONDecodeError:
+            # We do not care much about user scope data as it maybe blank
+            # so return blank dictionary instead
+            return {}
+
+    def get_access_token_data(self, request, app, client):
+        """ We need to gather the info from the apple specific login """
+        add_apple_session(request)
+
+        # Exchange `code`
+        code = get_request_param(request, "code")
+        access_token_data = client.get_access_token(code)
+
+        return {
+            **access_token_data,
+            **self.get_user_scope_data(request),
```

Comment:
I think we should keep the unverified `user` scope data in separate key so that it can does not overwrite the `access_token_data`. The `user` data can be altered from the client

Reply:
Ah, yes. Well, 2 things.

1. We _should_ whitelist the keys from `user` so that we know we're not passing in bad content.
2. The actual processing of the keys should depend on the signed id_token being from apple as expected. This does mean that any auth attempt will bail around this point if the request is invalid.